### PR TITLE
Factoring the API user authentication code

### PR DESCRIPTION
**Pull request by lasryaric**

### CODE REVIEW

File: front/lib/error.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+export type ErrorType =
+  | "invalid_request_error"
+  | "internal_server_error"
+  | "bad_request"
```
Comment (lasryaric) [inline front/lib/error.ts:4]: Errors to be refined

Comment (spolu) [reply front/lib/error.ts:4]: What do you mean by "refined" ? the list is incomplete now?

File: front/pages/api/v1/apps/[user]/[sId]/runs/index.ts [hunk 2]
```diff
@@ -37,81 +49,34 @@ const poll = async ({
...
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  const auth_result = await auth_api_user(req);
+  if (auth_result.isErr()) {
+    const err = auth_result.error();
+    return res.status(err.status_code).json(err.error);
   }
+  const authUser = auth_result.value();
```
Comment (spolu) [inline front/pages/api/v1/apps/[user]/[sId]/runs/index.ts:61]: authRes

File: front/lib/api/error.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  | "invalid_api_key_error"
+  | "internal_server_error";
+
+type _HTTPError = {
+  type: ErrorType;
+  message: string;
+  data_source_error?: object;
+};
+
+export type HTTPError = {
```
Comment (spolu) [inline front/lib/api/error.ts:13]: Rename APIError ?

File: front/pages/api/v1/apps/[user]/[sId]/runs/index.ts [hunk 4]
```diff
@@ -37,81 +49,34 @@ const poll = async ({
...
-        type: "invalid_api_key_error",
-        message: "The API key provided is invalid or disabled.",
-      },
-    });
-    return;
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  const authRes = await auth_api_user(req);
```
Comment (spolu) [inline front/pages/api/v1/apps/[user]/[sId]/runs/index.ts:56]: nit: preserve the parallelism here
```
 let [authRes, appOwner] = await Promise.all([
 auth_api_user(req),
 User.findOne({
    where: {
      username: req.query.user,
    },
  })
]);
```

File: front/pages/api/v1/apps/[user]/[sId]/runs/index.ts [hunk 5]
```diff
@@ -126,13 +91,13 @@ export default async function handler(req, res) {
...
       where: {
-        userId: reqUser.id,
+        userId: appOwner.id,
         sId: req.query.sId,
       },
     }),
     Provider.findAll({
       where: {
-        userId: reqUser.id,
+        userId: appOwner.id,
```
Comment (spolu) [inline front/pages/api/v1/apps/[user]/[sId]/runs/index.ts:100]: providers should always come from authUser (even if they are the same here)

File: front/pages/api/v1/apps/[user]/[sId]/runs/index.ts [hunk 6]
```diff
@@ -195,7 +160,7 @@ export default async function handler(req, res) {
             method: "POST",
             headers: {
               "Content-Type": "application/json",
-              "X-Dust-User-Id": appUser.id,
+              "X-Dust-User-Id": appOwner.id.toString(),
```
Comment (spolu) [inline front/pages/api/v1/apps/[user]/[sId]/runs/index.ts:163]: Should be `authUser` here (even if not yet in use). Was incorrect in the original file :+1:

File: front/pages/api/v1/data_sources/[user]/[name]/documents/[documentId]/index.ts [hunk 7]
```diff
@@ -215,7 +182,7 @@ export default async function handler(req, res) {
       }
 
       // Enforce FreePlan limit: 32 documents per DataSource.
-      if (reqUser.username !== "spolu") {
+      if (appOwner.username !== "spolu") {
```
Comment (spolu) [inline front/pages/api/v1/data_sources/[user]/[name]/documents/[documentId]/index.ts:185]: out of consistency even if will be delurked real soon, should be authUser

**Review by spolu — COMMENTED** [review]
Left minor comments. Looks great!

File: front/pages/api/v1/apps/[user]/[sId]/runs/index.ts (hunk 5 above)
Comment (lasryaric) [reply front/pages/api/v1/apps/[user]/[sId]/runs/index.ts:100]: Got it now 👍 But I was blindly replacing.

File: front/pages/api/v1/apps/[user]/[sId]/runs/index.ts (hunk 6 above)
Comment (lasryaric) [reply front/pages/api/v1/apps/[user]/[sId]/runs/index.ts:163]: 👍

File: front/pages/api/v1/data_sources/[user]/[name]/documents/[documentId]/index.ts [hunk 8]
```diff
@@ -169,7 +136,7 @@ export default async function handler(req, res) {
       let [providers] = await Promise.all([
         Provider.findAll({
           where: {
-            userId: reqUser.id,
+            userId: appOwner.id,
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply front/pages/api/v1/data_sources/[user]/[name]/documents/[documentId]/index.ts]: 👍

File: front/pages/api/v1/data_sources/[user]/[name]/documents/[documentId]/index.ts (hunk 7 above)
Comment (lasryaric) [reply front/pages/api/v1/data_sources/[user]/[name]/documents/[documentId]/index.ts]: 👍

File: front/lib/api/error.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+export type ErrorType =
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply front/lib/api/error.ts]: done

File: front/lib/api/error.ts (hunk 3 above)
Comment (lasryaric) [reply front/lib/api/error.ts]: done

**Review by spolu — APPROVED** [review]
LGTM

**Comment by spolu** [comment]
Very nice :ship:

**Pull request merged**